spec/app/ics-030-middleware/BEST_PRACTICES.md

```diff
@@ -0,0 +1,71 @@
+
+```json
+{
+    "<middleware_name>": "<middleware_version_value>",
+    "app_version": "<application_version_value>",
+}
+```
+
+The `<middleware_name>` key in the JSON struct should be replaced by the actual name of the key for the corresponding middleware (e.g. `fee` for ICS-29 fee middleware).
+
+In the application callbacks, the middleware can unmarshal the version string and retrieve the middleware and application versions. It must do its own version negotiation on `<middleware_version_value>` and then hand over `<application_version_value>` to the nested application's callback. This is only relevant if the middleware expects a compatible counterparty middleware at the same level on the counterparty stack. Middleware that only executes on a single side of the channel MUST NOT modify the channel version.
+
+## Packet Data Structuring For Middleware
+
+Packet senders may choose to send input data not just to the base application but also the middlewares in the stack. The packet data should be structured in the same way that the middleware stack is; i.e. nested from the top level middleware to the base application.
+
+Similar to the version negotiation, the `app_packet_data` may be marshalling packet data for underlying middleware as well.
+
+```json
+{
+    "<middleware_name>": "<middleware_packet_data>",
+    "app_packet_data": "<application_packet_data>",
+}
+```
+
+On `SendPacket`, the execution flow goes from the base application to the top layer middleware. Thus at each step, the middleware can choose to wrap the application packet data with its own packet data before returning to the `ICS4Wrapper`.
+
+On `ReceivePacket`, the middleware must process its own packet data, and then pass the `app_packet_data` to the underlying application. It may also choose to use the `app_packet_data` as part of its logic execution if it knows how to unmarshal the underlying data.
+
+Since the `OnRecvPacket` callback expects the full packet to be passed in, the packet data must be mutated like so:
+
+```typescript
+packetData = unmarshal(packet.data)
+
+// do whatever logic with middleware packet data
+handleMiddlewareData(packetData.MiddlewareData)
+
+// reassign packet data bytes before passing packet to next handler
+packet.Data = packetData.AppPacketData
+
+return im.app.OnRecvPacket(ctx, packet, relayer)
```

Comment:
Is this intended to handle arbitrary nesting of packet data when provided the full packet?